### PR TITLE
object.castShadowSide

### PR DESCRIPTION
**Pull request by tschw**
See https://github.com/mrdoob/three.js/commit/b07f8b94d4ad0ed915596c40f25c5e2584558e17#commitcomment-17315697

Related: #8771

### CODE REVIEW

File: src/core/Object3D.js [hunk 1]
```diff
@@ -579,7 +580,14 @@ Object.assign( THREE.Object3D.prototype, THREE.EventDispatcher.prototype, {
 
 		if ( this.name !== '' ) object.name = this.name;
 		if ( JSON.stringify( this.userData ) !== '{}' ) object.userData = this.userData;
-		if ( this.castShadow === true ) object.castShadow = true;
+
+		if ( this.castShadow === true ) {
+
+			object.castShadow = true;
+			object.castShadowSize = this.castShadowSide;
```
Comment (mrdoob) [inline src/core/Object3D.js]: Typo: `object.castShadowSize` (*Size instead of *Side)

Comment (mrdoob) [inline src/core/Object3D.js]: Better like this actually:

```js
if ( this.castShadow === true ) object.castShadow = true;
if ( this.castShadowSide !== THREE.BackSide ) object.castShadowSize = this.castShadowSide;
```

The user may have disabled `castShadow` temporarily.

File: src/loaders/ObjectLoader.js [hunk 2]
```diff
@@ -558,7 +558,15 @@ THREE.ObjectLoader.prototype = {
 
 			}
 
-			if ( data.castShadow !== undefined ) object.castShadow = data.castShadow;
```
Comment (mrdoob) [inline src/loaders/ObjectLoader.js]: Same here...

```js
if ( data.castShadow !== undefined ) object.castShadow = data.castShadow;
if ( data.castShadowSide !== undefined ) object.castShadowSide = data.castShadowSide;
```

File: src/core/Object3D.js (hunk 1 above)
Comment (WestLangley) [inline src/core/Object3D.js]: object.castShadowSide

**Comment by tschw** [comment]
@mrdoob 
Changed it in-place according to your comments. Also fixed that "castShadowSize" typo.

**Comment by WestLangley** [comment]
Thanks. I agree this is an improved API.

I am concerned, however, that it allows for options that do not make sense. For example.
```javascript
material.side = THREE.FrontSide;
object.castShadowSide = THREE.DoubleSide;
```

There are 3 X 3 = 9 combinations.

Clearly we want to support shadows on closed meshes (e.g., solid ones). That works by default.
```javascript
material.side = THREE.FrontSide; // default
object.castShadowSide = THREE.BackSide; // default
```

I assume we would like to support shadows on open meshes, too. Examples would be the Ninja mask (which is open at the bottom) and the cloth example. In both cases, we set `material.side = THREE.DoubleSide`, and I assume we want shadows to be cast from both front an back faces in those two examples.

This API supports that:
```javascript
material.side = THREE.DoubleSide;
object.castShadowSide = THREE.DoubleSide;
```

However, do any of the remaining 7 combinations make sense?

**Comment by tschw** [comment]
@WestLangley 
> I am concerned, however, that it allows for options that do not make sense.
> [...] do any of the remaining 7 combinations make sense?

What do we know what folks will come up with...? Or more specifically, why should we even care?

If we can agree on that it's sufficiently simple to use, let's just go with it: Just because a rifle can be used to deliberately shoot yourself in the foot, doesn't make it any less useful for shooting deer.

There are always the kind of people who stick wet dogs into microwave ovens, but those are immune to safety measures you build into software just as the tricky bugs are immune to automated testing. And I bet there's a similar pesticide effect; the harder you try the worse it gets :).

**Comment by WestLangley** [comment]
I think three.js should properly cast shadows cast from (1) `FrontSided` solid meshes and (2) `DoubleSided` open meshes. That's it. The library should just work.

If there are artifacts from self-shadowing, the user should be advised to adjust the shadow `bias`.

I now do not see the need for `object.castShadowSide`, which adds unneeded complexity.

Furthermore, since we support `MultiMaterial`, an an object can have many materials -- not all of which may be compatible with a single `object.castShadowSide` value.

**Comment by mrdoob** [comment]
@WestLangley So you think that `WebGLShadowMap` should basically do this?

```js
var map = {
    THREE.Frontside: THREE.BackSide,
    THREE.BackSide: THREE.FrontSide,
    THREE.DoubleSide: THREE.DoubleSide
};
result.side = map[ material.side ];
```

**Comment by WestLangley** [comment]
@mrdoob Actually, I do not care that much how it is implemented. I am focusing on the API and the user experience. That has been my concern through this entire discussion.

I propose that we ensure that shadows "just work" with front-sided closed meshes and double-sided open meshes. For back-sided meshes, I don't care if we support shadows or not.

**Comment by tschw** [comment]
@WestLangley 
> I propose that we ensure that shadows "just work" with front-sided closed meshes and double-sided open meshes.

That is, double-sided "just works" = without artefacts. dropping one side?
Or "just works" = being rendered completely and user will have to deal with the artefacts?

The former is currently implemented on `dev` and in this PR, the latter would require us to default `shadowMap.allowDoubleSided = true`.

I'm fine with (some kind of) automatism, which this one would remove entirely. However, I think there should be an off-switch somewhere in the lib for [several reasons](https://github.com/mrdoob/three.js/commit/b07f8b94d4ad0ed915596c40f25c5e2584558e17#commitcomment-17317249).

The off-switch is currently `shadowMap.flipSidedFaces = false, shadowMap.allowDoubleSided = true`.

> For back-sided meshes, I don't care if we support shadows or not.

Those are currently handled correctly and would need two properties to be set with the API in this PR.

**Comment by mrdoob** [comment]
@WestLangley 

> I now do not see the need for `object.castShadowSide`, which adds unneeded complexity.

We already had such complexity though:

```js
renderer.shadowMap.cullFace = THREE.CullFaceBack;
```

Some examples rely on it. After this PR, we would have basically moved `shadowMap.cullFace` to `object.shadowMapSide`. Which I think it's a nicer API.

**Comment by tschw** [comment]
@mrdoob 
> After this PR, we would have basically moved shadowMap.cullFace to object.shadowMapSide. Which I think it's a nicer API.

Now that I see it implemented, and after reading West's concerns, I'm actually not so sure I like this one better. I think what's currently implemented is, because it's automated and "just works". Renaming the settings or better defaults may help you like it better. I'll make another PR updating the examples to use `shadowMap.flipSidedFaces = false`.

**Comment by tschw** [comment]
Summary of the arguments that convinced me to drop this PR:

- There can be an N:M mapping between objects and materials (`MultiMaterial`).
- Fewer cases work OOTB.
- Typical use (other than going with the default) always requires an extra property to be set.
- That extra property resides in the data model.

**Comment by WestLangley** [comment]
>That is, double-sided "just works" = without artefacts. dropping one side?
Or "just works" = being rendered completely and user will have to deal with the artefacts?

The 2nd one. It has been my experience that setting `bias` has adequately resolved the artifacts in every use case I have encountered.

**Comment by tschw** [comment]
@WestLangley 
Thanks.

Then `allowDoubleSided` can be removed.

I fixed a problem with the depth codec in #8745, so maybe previous attempts of finding a suitable default panning value failed just because of that?

Further, it should now be possible now to apply the bias at the side of the encoder. This way, we could easily ensure a non-zero default when there are faces that both cast and receive shadows and keep the rest as-is for now. Once it's field-tested could optionally move away from inverted culling, if we want that. The approach would also save a uniform where they are scarce and moves it to where there's more headroom.

**Comment by mrdoob** [comment]
> Then `allowDoubleSided` can be removed.

👍

**Comment by WestLangley** [comment]
>Then allowDoubleSided can be removed.

Yes, please.

**Comment by tschw** [comment]
@mrdoob 
@WestLangley 

How to best call the remaining flag?
- `inverFaceCulling`
- `inverFaceCullingDirection`
- `reverseFaceCulling`
- `flipSidedFaces`
- `useFlipSidedFaces`
- `renderFlipSidedFaces`
- `oppositeSingleSidedFaces`
- `useOppositeSingleSidedFaces`
- `renderOppositeSingleSidedFaces`

...

Other than `.cullFace`, none of these need you to explain the specialities of the renderer to be understood.

**Comment by WestLangley** [comment]
The shadow mapping technique we are supposedly implementing is one in which we cull front faces by default. We treat double-sided materials as back-to-back faces.

Users do not need to know the details of how we are implementing this technique.

I prefer the API we have had all along: `shadow.map.cullFace = THREE.CullFaceFront`.

Hmmm.. [For awhile, now.](https://github.com/mrdoob/three.js/issues/2683#issuecomment-10723231)

**Comment by tschw** [comment]
@WestLangley 
> The shadow mapping technique we are supposedly implementing is one in which we cull front faces by default. 

"Culling front faces" actually means they are **not renderered.** This is standard terminology in computer graphics (and I hope we can agree on that part).

**But when** `material.side = THREE.FrontSide` **we do  render the back faces instead - thus they are actually *not* culled, but the other side *is* being rendered! It's also been that way before I made any changes, and also makes more sense than *actually* culling (i.E. *not rendering*) them.**

> Users do not need to know the details of how we are implementing this technique.

I totally agree on that point.

BUT: Consequently users should not be thinking of it as culling, at least not the way `.cullFace` suggests because it's very much misleading, as pointed out above.

Whether culling is used to invert the faces that are being rendered or other means is unimportant. 
The important bit is that the other side *is* being rendered! IMHO it makes a lot more sense to let the user specify that.

> I prefer the API we have had all along: shadow.map.cullFace = THREE.CullFaceFront.

I'm sorry to say, but regardless of "the shadow mapping technique we are supposedly implementing", this API is a very bad packaging for what's actually going on.

I gave you the commit to revert, so you can make it a PR and I'd feel very sorry for Three.js if @mrdoob actually merged it, because it would be a step back into obscurity.

I'm under the impression you might have interested motives, or, to state it most clearly, want to be the wise guy who knows secret semantics hidden behind an obscure interfaces. I actually do believe you can do without that, but I also don't think it's what software design, and OSS in particular should be about.

> We treat double-sided materials as back-to-back faces.

Can you elaborate on the double-sided case, isn't that contradicting to what you said [above](https://github.com/mrdoob/three.js/pull/8779#issuecomment-216331487)? :confused:

**Comment by mrdoob** [comment]
How about also removing `flipSidedFaces`? So `ShadowMap` will just render what the material says. Then we'll simplify all this to having to mess with `bias` only.

**Comment by tschw** [comment]
@mrdoob 
West's argumentation is contradictory and not making any sense to a degree I'm tempted to call "sabotage" :).

My above post and the current interface OTOH *do* make sense.

So *please* let's just find better names for the flags and be happy.

It is completely unimportant whether there's an article out there that suggests using the culling state for choosing particular faces to render. It won't make `.cullFace` a more intuitive API since it subverts the meaning of standard computer graphics terminology (as pointed out in my previous post).

Whatever way you use to get the sought faces rendered or not can't change a shadowmapping technique either - only rendering something different actually can.

I also had a look at the repository state at that time of the point of its origin

@tschw wrote:
> ... **they are actually *not* culled, but the other side *is* being rendered! It's also been that way before I made any changes**

...and it's **always** been that way! The handling of `material.side` has not changed

https://github.com/alteredq/three.js/blob/2fb5db39f896a5902cb47d6548f5cd12dc1ec69e/src/renderers/WebGLRenderer.js#L5979-L6016

So that API has always been as misleading as I found it. History is no excuse and straightening it out was long overdue.

**Comment by mrdoob** [comment]
Yes yes, I understand. Don't get lost in West words 😄

I'm actually proposing removing both flags altogether. [As you know](https://github.com/mrdoob/three.js/pull/8784/files), a bunch of example disable the "flipSiding" altogether.

So I propose removing both `flipSidedFaces` (aka `cullFace`) and `allowDoubleSided` and focus on `shadow.bias` instead.

**Comment by tschw** [comment]
Without setting a default that will cause a storm of support requests. The value of `-0.001` used by West is pretty close to the minimum `exp2(-10 ) == 0.0009...` which I measured (at both sides of the depth codec) and would have to be tested with reduced precision. Chrome has a start option to simulate that.

What's been holding me back to suggest it, is that this value seems still fairly large, and it's a fraction of the `near` ... `far` range, actually.

Ben also advocated the use of the back faces for shadowing because it's free of these pains (and he doesn't seem to care about anything but front-facing materials, since everything can be modeled as a solid body).

What we have in `dev` right now, works in all cases OOTB. When you want double sided materials to cast a shadow from both sides or cast shadows of front faces you have to set a flag, respectively.

When accept THREE.BackSide to always require panning, we could simplify it to `forceBackFaces` (enabled by default).

**Comment by mrdoob** [comment]
Thanks! That cleared things up for me. 

So, how about `.renderBackFaces`?

**Comment by tschw** [comment]
@mrdoob 
> That cleared things up for me. So, how about shadowMap.renderBackFaces true by default? 

I'd personally find it better to just keep the `THREE.BackSide` case working.

Here's a suggestion of a compromise that will keep the idea of culling something, it also keeps the "just works" behavior in all cases, and immediately makes sense to anyone familiar with standard terms:

`.invertFaceCulling`

*Whether to render the opposite side as specified by the material into the shadow map. Default is `true`. When set to `false` an appropriate shadow bias must be set on the `Light`.*

`.forceFaceCulling`

*Whether to treat materials specified as double-sided as if they were specified as front-sided when rendering the shadow map. Default is `true`. When set to `false` an appropriate shadow bias must be set on the `Light`.*

**Comment by mrdoob** [comment]
Ok. These sound good to me! 👍

**Comment by mrdoob** [comment]
However, if we wanted to get away from "culling". A less intimidating naming could be `.renderBackSide` (default `true`) and `.renderSingleSide` (default `true`).

**Comment by tschw** [comment]
@mrdoob 
> Ok. These sound good to me! :+1:

Great! You like them - I went with those :)

> A less intimidating naming could be `.renderBackSide`

I don't find the "culling" part intimidating, but the way it was specified via `cullFace` in combination with the existence of a `material.side`: It suggests that something gets excluded from rendering which is in fact just rendered inside-out.

`renderBackSide` has a bit of a similar problem: Although a lot easier grasp than `.cullFace` (which is of course a very subjective statement), the "back side" we are talking about is still "the back side from the perspective of the light after applying the sidedness of the material" unless breaking the `THREE.BackSide` case deliberately in favor of simplicity of specification.

Gosh, what a beast of a case... :sweat:

**Comment by mrdoob** [comment]
`renderReverseSide`? 😇

**Comment by tschw** [comment]
@mrdoob 
`.renderReverseSide`
`.renderSingleSide`

Basically like those. I think we should either append an "s" (plural) or a "d" (passive). Singular may yield the conclusion it would be about the entire map rather than its contents, which wouldn't make any sense, of course, but still, the reader may or may not figure that.

**Comment by mrdoob** [comment]
I think "d".

**Comment by WestLangley** [comment]
@tschw said
>But when material.side = THREE.FrontSide we do render the back faces instead - thus they are actually not culled, but the other side is being rendered! 

To me, there is no such concept as rendering the "other side of a face" -- especially when rendering to a depth map.

One can render _different_ faces, however, by _culling_ different faces. That is what is happening.

The renderer sets the front-face winding order and culls faces when rendering with `MeshDepthMaterial` -- just as it does it when rendering with any other material.

The so-called "flipping" of faces was the chosen technique to get the front faces culled instead of the back ones.

@tschw said
>I'm under the impression you might have interested motives, or, to state it most clearly, want to be the wise guy who knows secret semantics hidden behind an obscure interface.

@tschw I have treated you respectfully. Based on the content of this thread (and earlier ones), you apparently do not feel the same obligation to treat me respectfully -- especially if I express an opinion that you do not like.

Personally, I am tired of it.

You asked for my opinion, and I gave it. I expect you to afford me the same respect I afford you. I do not expect personal attacks in response.

**Comment by mrdoob** [comment]
Peace guys 😢

**Comment by tschw** [comment]
@WestLangley 
I think you can be brilliant when you're making sense. In fact, I closed this very PR because you expressed concerns that convinced me it was the wrong way to go. How is that *not* respecting you?

However, your overall criticism has been rather destructive throughout these PRs and I've really been wondering why. Deliberately non getting a very clear point, and instead presenting nonsensical, though-terminating and contradictory arguments that don't relate to the presented line of argumentation, just to tell me in the very end that you want the old interface back and didn't even bother to consider anything else, is not very productive and raises questions. Why don't you say so right from the start and waste all of our time? It's just as potentially offending as you find my statement above and any reader of this discussion may easily arrive at a similar conclusion about your motives.

There's no personal disrespect involved from my side, just maybe a bit of an invitation for self-reflection on how your actions will be perceived.

I read up on the history and I see that you brought it into existence with best intentions. I also think it was reasonable within a certain scope. Maybe it just felt so right back then, that you had to defend it here.

HTH and won't make it worse :).

For your technical arguments, I *do* get what you are saying: 

As said ([here](https://github.com/mrdoob/three.js/pull/8779#issuecomment-216895483)), I even agree that `.cullFace` would be OK without a `material.side` in the API. With it, the whole thing becomes confusing and difficult to explain. You have just demonstrated it IMO.
It's also insufficient when we want all settings to "just work". I actually *did* consider keeping it, but when you think about it a little longer, I'm sure you'll also notice that it won't sufficiently capture all the cases.

**Pull request closed without merging**